[AIRFLOW-247] Add EMR hook, operators and sensors. Add AWS base hook

rfroetscher opened the pull request:
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- https://issues.apache.org/jira/browse/AIRFLOW-247

Testing Done:
- Unit test for each added module
- Manual testing of each example dag included using our own AWS account

Open questions:
- Does it make sense for the separate operators to be somehow combined or to inherit from a common parent? They share some AWS response checking logic, but it felt like inheriting from a common parent would only reduce that one small bit of duplication while introducing some confusing complexity.
- Currently for `EmrCreateJobFlowOperator` job_flow_overrides are a python dictionary that allow DAG specific overrides for the EMR connection config (contained in the extra of an Airlfow Connection). However, this python dictionary doesn't allow for templating (for example if you wanted to name your cluster differently for each execution date). Would it make more sense for it to be json overrides, so they could be templated?

rfroetscher committed:
[AIRFLOW-247] Add EMR hook, operators and sensors. Add AWS base hook

artwr commented on airflow/contrib/operators/emr_create_job_flow_operator.py:
```diff
@@ -0,0 +1,61 @@
+            self,
+            aws_conn_id='s3_default',
+            emr_conn_id='emr_default',
+            job_flow_overrides={},
```
A better practice would be to have this as `None`, and override it with an empty dictionary inside the `init`. See http://reinout.vanrees.org/weblog/2012/04/18/default-parameters.html for instance.

rfroetscher replied:
hmm, that's good to know, thanks. i'm pretty new to python so it's good to learn its quirks.

rfroetscher committed:
job_flow_overrides defaults to None, in init proper dictionary defaulting takes place

in python, you can unintentionally be pointing to the same dictionary in different usages of kwarg defaulting.
http://reinout.vanrees.org/weblog/2012/04/18/default-parameters.html
